src/Digdir.Domain.Dialogporten.Application/Common/IDialogTokenGenerator.cs

```diff
@@ -0,0 +1,135 @@
+    // so having the token being invalid after 30 minutes should be sufficient.
+    private readonly TimeSpan _tokenLifetime = TimeSpan.FromMinutes(30);
+
+    public DialogTokenGenerator(
+        IOptions<ApplicationSettings> applicationSettings,
+        IUser user,
+        ICompactJwsGenerator compactJwsGenerator)
+    {
+        _applicationSettings = applicationSettings.Value ?? throw new ArgumentNullException(nameof(applicationSettings));
+        _user = user ?? throw new ArgumentNullException(nameof(user));
+        _compactJwsGenerator = compactJwsGenerator ?? throw new ArgumentNullException(nameof(compactJwsGenerator));
+    }
+
+    public string GetDialogToken(DialogEntity dialog, DialogDetailsAuthorizationResult authorizationResult)
+    {
+        var claimsPrincipal = _user.GetPrincipal();
+        var now = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
+        var dt = new DialogTokenClaims
+        {
+            JwtId = Guid.NewGuid(),
+            DialogId = dialog.Id,
+            AuthenticatedParty = GetAuthenticatedParty(),
+            AuthenticationLevel = claimsPrincipal.TryGetAuthenticationLevel(out var authenticationLevel)
+                ? authenticationLevel.Value
+                : 0,
+            DialogParty = dialog.Party,
+            SupplierParty = claimsPrincipal.TryGetSupplierOrgNumber(out var supplierOrgNumber)
+                ? NorwegianOrganizationIdentifier.PrefixWithSeparator + supplierOrgNumber
+                : null,
+            Actions = GetAuthorizedActions(authorizationResult),
+            Issuer = _applicationSettings.Dialogporten.BaseUri.ToString(),
+            IssuedAt = now,
+            NotBefore = now,
+            Expires = now + (long)_tokenLifetime.TotalSeconds
+        };
+
+        return _compactJwsGenerator.GetCompactJws(dt);
+    }
+
+    private static string GetAuthorizedActions(DialogDetailsAuthorizationResult authorizationResult)
+    {
+        var actions = new StringBuilder();
+        foreach (var (action, resource) in authorizationResult.AuthorizedAltinnActions)
+        {
+            actions.Append(action);
+            if (resource != Authorization.Constants.MainResource)
+            {
+                actions.Append(CultureInfo.InvariantCulture, $",{resource}");
+            }
+
+            actions.Append(';');
+        }
+
+        // Remove trailing semicolon
+        actions.Remove(actions.Length - 1, 1);
+
+        return actions.ToString();
+    }
+
+    private string GetAuthenticatedParty()
+    {
+        if (_user.TryGetPid(out var pid))
+        {
+            return NorwegianPersonIdentifier.PrefixWithSeparator + pid;
+        }
+
+        if (_user.TryGetOrgNumber(out var orgNumber))
+        {
+            return NorwegianOrganizationIdentifier.PrefixWithSeparator + orgNumber;
+        }
+
+        throw new InvalidOperationException("User must have either pid or org number");
+    }
+}
+
+// TODO! Consider moving this to the domain layer
```

Comment:
Bør dette til Domain? Brukes av IDialogTokenGenerator og ICompactJwsGenerator

Reply:
Kan nok argumentere for at dette er versjonsspesifikt. Kanskje ICompactJwsGenerator i stedet burde ta en `IEnumerable<Claim>`?

Reply:
Endret til å manuelt bygge en `Dictionary<string, object?>` i stedet. (`Claim` forutsetter string-verdier, som blir feil for `exp`, `iat` og `nbf` jf. standard)